llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -0,0 +1,427 @@
+  // mark all global variables as needed and remove dead ones after cloning.
+  // Notice. For device global variables with the 'device_image_scope' property,
+  // removing dead ones is a must, the 'checkImageScopedDeviceGlobals' function
+  // checks that there are no usages of a single device global variable with the
+  // 'device_image_scope' property from multiple modules and the splitter must
+  // not add such usages after the check.
+  for (const auto &G : M.globals())
+    GVs.insert(&G);
+}
+
+ModuleDesc extractSubModule(const Module &M,
+                            const SetVector<const GlobalValue *> GVs,
+                            const EntryPointGroup &ModuleEntryPoints) {
+  ModuleDesc Res{nullptr, ModuleEntryPoints};
+
+  // For each group of entry points collect all dependencies.
+  ValueToValueMapTy VMap;
+  // Clone definitions only for needed globals. Others will be added as
+  // declarations and removed later.
+  Res.M = CloneModule(M, VMap,
+                      [&](const GlobalValue *GV) { return GVs.count(GV); });
+
+  EntryPointVec &NewEPs = Res.EntryPoints.Functions;
+  // replace entry points with cloned ones
+  std::for_each(NewEPs.begin(), NewEPs.end(),
+                [&VMap](const Function *F) { return cast<Function>(VMap[F]); });
+
+  return Res;
+}
+
+// TODO: try to move including all passes (cleanup, spec consts, compile time
+// properties) in one place and execute MPM.run() only once.
+void cleanupSplitModule(Module &SplitM) {
+  ModuleAnalysisManager MAM;
+  MAM.registerPass([&] { return PassInstrumentationAnalysis(); });
+  ModulePassManager MPM;
+  // Do cleanup.
+  MPM.addPass(GlobalDCEPass());           // Delete unreachable globals.
+  MPM.addPass(StripDeadDebugInfoPass());  // Remove dead debug info.
+  MPM.addPass(StripDeadPrototypesPass()); // Remove dead func decls.
+  MPM.run(SplitM, MAM);
+}
+
+// The function produces a copy of input LLVM IR module M with only those entry
+// points that are specified in ModuleEntryPoints vector.
+ModuleDesc extractCallGraph(const Module &M,
+                            const EntryPointGroup &ModuleEntryPoints) {
+  SetVector<const GlobalValue *> GVs;
+  collectFunctionsToExtract(GVs, ModuleEntryPoints);
+  collectGlobalVarsToExtract(GVs, M);
+
+  ModuleDesc SplitM = extractSubModule(M, GVs, ModuleEntryPoints);
+  cleanupSplitModule(SplitM.getModule());
+
+  return SplitM;
+}
+
+class ModuleCopier : public ModuleSplitterBase {
+public:
+  using ModuleSplitterBase::ModuleSplitterBase; // to inherit base constructors
+
+  ModuleDesc nextSplit() override {
+    return {releaseInputModule(), nextGroup()};
+  }
+};
+
+class ModuleSplitter : public ModuleSplitterBase {
+public:
+  using ModuleSplitterBase::ModuleSplitterBase; // to inherit base constructors
+
+  ModuleDesc nextSplit() override {
+    return extractCallGraph(getInputModule(), nextGroup());
+  }
+};
+
+} // namespace
+
+bool llvm::module_split::ModuleDesc::isEsimd() {
```

Comment:
Just a nit: it looks like the `llvm` namespace prefix can be removed here too in the same way as you did it for the `module_split::getSlitterByKernelType` function.